### PR TITLE
Support Bss StoreCredit plugin

### PR DESCRIPTION
**Pull request by oldDriver**
# Description
Support Bss StoreCredit plugin

Fixes: https://app.asana.com/0/0/1142368638742826/f

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [x] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [ ] I have created or modified unit tests to sufficiently cover my changes.

### CODE REVIEW

**Comment by matt-thomason** [comment]
@oldDriver I am not referring to your test phrase, I'm asking if you're actually tested it with a merchant's BSS storecredit plugin. This does not look like it would work to me.

**Comment by oldDriver** [comment]
@matt-thomason i tested this with @eugene-shab on my local instance and on the Merchant's stage server

**Comment by danac-gs** [comment]
LGTM so far. @matt-thomason I believe that if Store Credit amount is added to the quote totals then this approach might actually work. In most of the cases. @oldDriver However, consider the case when the customer Store Credit total available balance is greater than the current shopping cart items subtotal. If store credit can be used to pay for shipping and / or tax then the total available amount should be passed to cart as a fixed_amount type discount. This can not be read from the session / quote / totals but you need to fetch it from some customer / store credit resource. Similar as the support for other store credit / reward point modules. If, on the other hand, this store credit can only be used to pay for the cart items, this solution may be sufficient.

**Comment by YevhenShBolt** [comment]
@matt-thomason ,
As i see all calculation of discount happen in

https://github.com/BoltApp/bolt-magento2/blob/ed59919460aefc2bbd4c2fe0890eeee51ad6729f/Helper/Cart.php#L1546

We always add new total type in such way (you can see it on example of AMASTY_GIFTCARD and etc). 

Previously we have a special method `$this->getDiscountTypes()`which returned a private variable  `$this->discountTypes` and have easy to extend through Magento plugin logic (see before plugin), but now we only left private variable.

So, in that way we can only adds items to the `discountTypes` array.

Also, often we add additional logic to the `Helper/Cart` as describe Daniel.

In this helper a lot of example like AMASTY_GIFTCARD, MAGEPLAZA_GIFTCARD, UNIRGY_GIFT_CERT and etc.

**Comment by danac-gs** [comment]
@YevhenShBolt @oldDriver please check how other Store Credit / Reward Point modules are supported as fixed_amount discounts:
https://github.com/BoltApp/bolt-magento2/blob/ed59919460aefc2bbd4c2fe0890eeee51ad6729f/Helper/Cart.php#L1428
https://github.com/BoltApp/bolt-magento2/blob/ed59919460aefc2bbd4c2fe0890eeee51ad6729f/Helper/Cart.php#L1469
https://github.com/BoltApp/bolt-magento2/blob/ed59919460aefc2bbd4c2fe0890eeee51ad6729f/Helper/Cart.php#L1503
To test your solution create a cart with subtotal less then a customer Store Credit balance and shipping and tax greater than 0.

**Comment by oldDriver** [comment]
Last changes are needed for Ambush\BuyWake

File: Helper/Discount.php [hunk 1]
```diff
@@ -49,6 +49,9 @@ class Discount extends AbstractHelper
     const MAGEPLAZA_GIFTCARD = 'gift_card';
     const MAGEPLAZA_GIFTCARD_QUOTE_KEY = 'mp_gift_cards';
     const AHEADWORKS_STORE_CREDIT = 'aw_store_credit';
+    const BSS_STORE_CREDIT = 'bss_storecredit';
+    const AMASTY_COUPON_AMOUNT = 'amasty_coupon_amount';
+    const AHEADWORKS_GIFTCARD = 'aw_giftcard';
```
Comment (roopakv) [inline Helper/Discount.php]: it seems like previously we already added `AHEADWORKS_STORE_CREDIT` is there anyway we can find all `aheadworks` discounts and add them now? same for amasty. we should just be able to whitelist another module once and for all as a whole?

Comment (danac-gs) [reply Helper/Discount.php]: @roopakv those are different modules of the same vendor. Support for a module can be a not so trivial task and so far it was done on demand (when a merchant needs it we build it into our plugin as an integral part).
@oldDriver did you test the case I explained above with all of the newly added discounts?

Comment (roopakv) [reply Helper/Discount.php]: @danac-gs thank you for explaining

Comment (oldDriver) [reply Helper/Discount.php]: @roopakv as for me better way to change property $discountTypes from "private" to "protected"
I this way we will not need to add all gift/discounts/coupons/store credit/voucher in the Helper\Cart.
We could simply add Botpay\Custom\Helper\Cart:
namespace Bolt\Custom\Helper
use Bolt\Boltpay\Helper\Cart as BaseCart;
class Cart extends BaseCart
{
    protected $discountTypes = [
        // Types we need in the current integration
    ];
}
Does it make sense?

Comment (roopakv) [reply Helper/Discount.php]: sort of. not following a 100% but i think i get the idea

**Comment by matt-thomason** [comment]
Closing this as it was addressed in #420

**Pull request closed without merging**